### ISSUE TITLE
OutOfMemoryError: {Java heap space, GC}

### ISSUE THREAD

**Issue by dr0i** [hbz/lobid#107]
The crashes take place at play lobid api instance.
Got an error which forced a restart:

    2015-01-19 14:48:19,476 - [ERROR] [...]
    java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3326) ~[na:1.8.0_05]
        at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137) ~[na:1.8.0_05]
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:121) ~[na:1.8.0_05]
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:421) ~[na:1.8.0_05]
        at java.lang.StringBuilder.append(StringBuilder.java:136) ~[na:1.8.0_05]
        at java.lang.StringBuilder.append(StringBuilder.java:76) ~[na:1.8.0_05]
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:457) ~[na:1.8.0_05]
        at java.lang.StringBuilder.append(StringBuilder.java:166) ~[na:1.8.0_05]
        at java.lang.StringBuilder.append(StringBuilder.java:76) ~[na:1.8.0_05]
        at com.google.common.base.Joiner.appendTo(Joiner.java:111) ~[com.google.guava.guava-16.0.1.jar:na]
        at com.google.common.base.Joiner.appendTo(Joiner.java:152) ~[com.google.guava.guava-16.0.1.jar:na]
        at com.google.common.base.Joiner.join(Joiner.java:193) ~[com.google.guava.guava-16.0.1.jar:na]
        at com.google.common.base.Joiner.join(Joiner.java:183) ~[com.google.guava.guava-16.0.1.jar:na]
        at controllers.Application.serialization(Application.java:292) ~[lobid.lobid-1.16.0.jar:1.16.0]

**Comment by dr0i** [comment]
Restart was necessary to solve the issue.

**Comment by dr0i** [comment]
Same issue again at 2015-01-20 10:51:41

**Comment by dr0i** [comment]
Again at 2015-05-20 09:43:55:

    2015-02-25 09:43:55,351 - [ERROR] - from akka.actor.ActorSystemImpl in play-akka.actor
.default-dispatcher-38109·
    Uncaught error from thread [play-akka.actor.default-dispatcher-38110] shutting down JV
    M since 'akka.jvm-exit-on-fatal-error' is enabled
    java.lang.OutOfMemoryError: GC overhead limit exceeded

**Comment by dr0i** [comment]
Again a memory problem at 2015-03-17 13:47:41,314 

    - [ERROR] - from akka.actor.ActorSystemImpl in play-akka.actor.default-dispatcher-1156  
    Uncaught error from thread [play-akka.actor.default-dispatcher-1132] shutting down JVM since   'akka.jvm-exit-on-fatal-error' is enabled
    java.lang.OutOfMemoryError: GC overhead limit exceeded

**Comment by philboeselager** [comment]
+1: 2015-03-24 09:33:46,808
+1: 2015-03-24 10:10:20,241

**Comment by philboeselager** [comment]
2015-03-24 10:10:20,241 - [ERROR] - from akka.actor.ActorSystemImpl in play-akka.actor.default-dispatcher-9 
Uncaught error from thread [play-akka.actor.default-dispatcher-3] shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.Arrays.copyOf(Arrays.java:3326) ~[na:1.8.0_05]
	at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137) ~[na:1.8.0_05]

**Comment by philboeselager** [comment]
There was a lot of log on the Play Console after having restarted between the last two crashes, coming at least in parts from nwbib.

**Comment by dr0i** [comment]
2015-05-29 09:47:48,071 
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.lang.StringCoding$StringEncoder.encode(StringCoding.java:300) ~[na:1.8.0_05]
	at java.lang.StringCoding.encode(StringCoding.java:344) ~[na:1.8.0_05]
	at java.lang.String.getBytes(String.java:906) ~[na:1.8.0_05]
	at play.api.mvc.Codec$$anonfun$javaSupported$1.apply(Results.scala:317) ~[com.typesafe.play.play_2.11-2.3.4.jar:2.3.4]

**Comment by dr0i** [comment]
Found the cause: it lies in the `play.cache.Cache` library. Setting a ttl like e.g. `Cache.set("a", "b", 1)` one would expect all cached objects would be removed after about one second, but that's not the case. Tried it with `https://github.com/hbz/lobid/blob/master/test/tests/SampleUsage.java` (which lookups 351k docs and repeatedly crashes the API).

So the easiest solution is to not use the cache functionality, at least in https://github.com/hbz/lobid/blob/master/app/models/Search.java - while it may be safer (and making more sense) to keep the cache in https://github.com/hbz/lobid/blob/master/app/controllers/Facets.java).

Will test manually  if the APIs behave  fine without the cache in `Search.java`.

**Comment by philboeselager** [comment]
Have you tried to alter the cache settings? The most obvious setting would be the cache persistence time, which is even optional. While having no knowledge about the memory consumption caused by the use of the classes you named, it seems to be possible that perhaps giving a shorter time than 1 day for Facets could help.

Searching the web didn't bring up many cases of memory leaks for the Play cache to me - but of course you never know. Found a good way to analyse heap dumps on http://stackoverflow.com/questions/17342760/memory-leak-in-play-framework-application .

**Comment by dr0i** [comment]
I profiled with the mentioned jprofiler (10 day trial version) and find also good enough the free http://www.eclipse.org/mat/ .Both tools showed that `ehcahe` is the problem.
Now, the forementioned `Cache.set("a", "b", 1)`should set `ttl` (Time To Live) to one second , which didn't worked. For the facets, it's ` 24 * 60 * 60` which is one day in seconds. Don't know if this works, but it doesn't matter, I think, because the used facets are not so many in numbers (i.e. not 350k in 3 hours).

In short: I think we shouldn't change the cache settings for facets while we don't need it in `Search`at all.

I don't know what we are missing tho make `ehcache` working as expected - it is widely used and I couldn't found a bug report. Yet, for our purpose we really *don't need* to be active.

**Comment by dr0i** [comment]
Deployed to staging and to production. Mind you that only lobid API is affected, NOT nwbib API (which has some other stuff done which I don't have evaluated yet, so that @fsteeg may have a look at it when necessary). For NWBib API look at hbz/nwbib#108.
Closing this one.
